src/display/xml_parser.js

```diff
@@ -0,0 +1,362 @@
+              let complexDoctype = false;
+              q = s.indexOf('>', j + 8);
+              if (q < 0) {
+                this.onError(XMLParserErrorCode.UnterminatedDoctypeDeclaration);
+                return;
+              }
+              if (q2 > 0 && q > q2) {
+                q = s.indexOf(']>', j + 8);
+                if (q < 0) {
+                  this.onError(
+                    XMLParserErrorCode.UnterminatedDoctypeDeclaration);
+                  return;
+                }
+                complexDoctype = true;
+              }
+              const doctypeContent =
+                s.substring(j + 8, q + (complexDoctype ? 1 : 0));
+              this.onDoctype(doctypeContent);
+              // XXX pull entities ?
+              j = q + (complexDoctype ? 2 : 1);
+            } else {
+              this.onError(XMLParserErrorCode.MalformedElement);
+              return;
+            }
+            break;
+          default:
+            const content = this._parseContent(s, j);
+            if (content === null) {
+              this.onError(XMLParserErrorCode.MalformedElement);
+              return;
+            }
+            let isClosed = false;
+            if (s.substring(j + content.parsed,
+                            j + content.parsed + 2) === '/>') {
+              isClosed = true;
+            } else if (s.substring(j + content.parsed,
+                       j + content.parsed + 1) !== '>') {
+              this.onError(XMLParserErrorCode.UnterminatedElement);
+              return;
+            }
+            this.onBeginElement(content.name, content.attributes, isClosed);
+            j += content.parsed + (isClosed ? 2 : 1);
+            break;
+        }
+      } else {
+        do {
+          /* skipping text items */
+        } while (j++ < s.length && s[j] !== '<');
+        const text = s.substring(i, j);
+        this.onText(this._resolveEntities(text));
+      }
+      i = j;
+    }
+  }
+
+  onPi(name, value) { }
+
+  onComment(text) { }
+
+  onCdata(text) { }
+
+  onDoctype(doctypeContent) { }
+
+  onText(text) { }
+
+  onBeginElement(name, attributes, isEmpty) { }
+
+  onEndElement(name) { }
+
+  onError(code) { }
+}
+
+class SimpleDOMNode {
+  constructor(nodeName, nodeValue) {
+    this.nodeName = nodeName;
+    this.nodeValue = nodeValue;
+
+    Object.defineProperty(this, 'parentNode', { value: null, writable: true, });
+  }
+
+  get firstChild() {
+    return this.childNodes[0];
+  }
+
+  get nextSibling() {
+    let index = this.parentNode.childNodes.indexOf(this);
+    return this.parentNode.childNodes[index + 1];
+  }
+
+  get textContent() {
+    if (!this.childNodes) {
+      return this.nodeValue || '';
+    }
+    return this.childNodes.map(function(child) {
+      return child.textContent;
+    }).join('');
+  }
+
+  hasChildNodes() {
+    return this.childNodes && this.childNodes.length > 0;
+  }
+}
+
+export class SimpleXMLParser extends XMLParserBase {
```

Comment:
Usually we put the exports at the bottom of the file. Can we do that here too for consistency (compare https://github.com/mozilla/pdf.js/pull/9573/files#diff-3cb21ac6c8ce15118704e14df3dcc9b2L405)?